tests/functional/test_bad_url.py

```diff
@@ -0,0 +1,23 @@
+# test the error message returned by pip when
+# a bad "file:" URL is passed to it.
+
+from typing import Any, Tuple
+
+
+def get_url_error_message(script: Any, fake_file: str) -> Tuple[str, str, int]:
+    # this makes pip to react using
+    # subprocess. It must fail, so then
+    # we can test the error message.
+    proc = script.pip("install", "-r", f"file:{fake_file}", expect_error=True)
+    expected_message = "ERROR: 404 Client Error: FileNotFoundError for url: "
+    return proc.stderr, expected_message, proc.returncode
```

Comment:
Do you plan to add more tests that use this function? If not, this entire function body should be moved into `test_filenotfound_error_message`.

Reply:
Good point. I am not planning to add more functions, so maybe I can follow your suggestion.